large screen support (WIP)

abcghy opened the pull request:
issue: #338 

Guideline: https://developer.android.com/jetpack/compose/layouts/adaptive
Official Reference: https://github.com/android/compose-samples/tree/main/JetNews
Design Reference: https://m3.material.io/foundations/layout/canonical-layouts/list-detail

JunkFood02 commented:
FYI: [Compose Material3 Adaptive Version](https://developer.android.com/jetpack/androidx/releases/compose-material3#compose_material3_adaptive_version_10_2)

mbestavros commented:
@Ashinch If delaying @JunkFood02's Compose update (#502) is holding community development back, why not merge it now? Especially if Greader needs a bit more time to cook as well.

Ashinch commented:
> @Ashinch If delaying @JunkFood02's Compose update (#502) is holding community development back, why not merge it now? Especially if Greader needs a bit more time to cook as well.

Hi, @mbestavros 

Google Reader API does indeed need some improvement time. Do you have any suggestions? For example, creating a new `dev` branch to merge these PRs, allowing the new PRs to continue development using the `dev` branch，after stabilizing, merge the `dev` branch back into the `main` branch? 🤔

Ashinch commented:
> > @Ashinch If delaying @JunkFood02's Compose update (#502) is holding community development back, why not merge it now? Especially if Greader needs a bit more time to cook as well.
> 
> Hi, @mbestavros 
> 
> Google Reader API does indeed need some improvement time. Do you have any suggestions? For example, creating a new `dev` branch to merge these PRs, allowing the new PRs to continue development using the `dev` branch，after stabilizing, merge the `dev` branch back into the `main` branch? 🤔

The reason for blocking #502 is that it involves a significant dependency upgrade, and I'm not sure if the UI will work properly. Currently, I'm dealing with issues related to the Google Reader API, so I might not have time to migrate the changes brought by #502.

JunkFood02 commented:
> not sure if the UI will work properly

Have been using the app with latest Compose for a while (not much!), and I didn't see any noticeable glitch or bug in it. To be honest, since stable release (v1.0) is still far ahead, users are more open to help test potential issues as long as there're still updates

Ashinch commented:
Hi, @abcghy 

Could you please help resolve a code conflict? I think some of our infrastructure is now ready.

Ashinch commented:
Hi, everyone.

What is the current status of this PR?

Has anyone else tried it yet?

I want to know if it can be merged. ☺️

JunkFood02 commented:
I've just took a look of this implementation before. It's rather a simple walkaround for large screen support, but lack of handling many corner cases, like rotating the screen or handling system navigation. Anyway, existing users(compact screen width) should not be affected by this change so i'm approving this change

JunkFood02 commented on app/src/main/java/me/ash/reader/infrastructure/android/MainActivity.kt:
```diff
@@ -82,7 +86,7 @@ class MainActivity : AppCompatActivity() {
                             }
                         }
-                        HomeEntry(subscribeViewModel = subscribeViewModel)
+                        HomeEntry(subscribeViewModel = subscribeViewModel, widthSizeClass = widthSizeClass)
```
Use CompositionLocal instead of a explicit parameter seems neater

JunkFood02 commented on app/src/main/java/me/ash/reader/ui/page/common/HomeEntry.kt:
```diff
@@ -152,13 +157,14 @@ fun HomeEntry(
             }
             forwardAndBackwardComposable(route = "${RouteName.READING}/{articleId}") {
-                ReadingPage(navController = navController, homeViewModel = homeViewModel)
+                ReadingPage(navController = navController, homeViewModel = homeViewModel, isExpandedScreen = isExpandedScreen)
```
TODO: handle system navigations & screen orientation changes

abcghy replied:
@JunkFood02  Could you elaborate on this one? Currently I only added a FlowRoute so that it can react to different width, that include the screen orientation changes.

Ashinch commented:
Okay, please remind me when this PR is ready.

abcghy commented:
1. added `isArticleOpen` and `articleId` in HomeUiState.
2. Combine Reading Page with Flow Page into FlowRoute. So that we can determine which page to show according to screen width and `isArticleOpen`.
3. Removed Reading Page from navigation due the reason above.

@JunkFood02

JunkFood02 commented on app/src/main/java/me/ash/reader/ui/page/home/flow/FlowRoute.kt:
```diff
@@ -0,0 +1,77 @@
+                onArticleClick = selectArticle,
+            )
+        }
+        FlowScreenType.ArticleDetails -> {
```
The `Article -> Flow -> Home` backstack seems broken now

JunkFood02 replied:
Anyway, we can put this change on hold since even Google haven't agree on the pattern of making a adaptive navigation

JunkFood02 commented:
Finally google stablize their large screen support library! We can begin to leverage `material3-adaptive` in our app

References:
https://www.youtube.com/watch?v=wBc6oMVAP60
https://developer.android.com/reference/kotlin/androidx/compose/material3/adaptive/package-summary
